### DISCUSSION TITLE
PrimeNG filter menu customization?

### DISCUSSION THREAD

**Discussion by vokic** [primefaces/community#626] [PrimeNG] [answered]
Is there a way to customize the context popup menu of a table filter, say to remove the "Add Rule" button, remove "Match" dropdown menu and set it to always be "Match any"? I can't find any configuration except filterMatchModeOptions

https://primeng.org/table#filter

![filter primeng](https://github.com/primefaces/community/assets/30930109/e417fde8-dca3-463f-a091-bdaf3712e057)

**Comment by ThoSap** [comment, marked as answer]
@cetincakiroglu that is simply wrong and I'm amazed that the PrimeNG team does not know that.
You can already do the things @vokic asked, it is just not documented anymore since PrimeNG v15 and I've already opened an issue for that months ago...
https://github.com/primefaces/primeng/issues/13434

@vokic you can find "all" the properties here in the PrimeNG v14 documentation.
https://www.primefaces.org/primeng-v14-lts/table#:~:text=ColumnFilter%20Properties

Some are missing in that documentation since I've opened a PR that was merged some time ago into PrimeNG 15.0.1 to add more of them to pass through to the custom ColumnFilter Template element `<p-columnFilterFormElement></p-columnFilterFormElement>` 
https://github.com/primefaces/primeng/pull/12312
https://github.com/primefaces/primeng/issues/11466

The things you asked can be done on the `<p-columnFilter></p-columnFilter>` element as follows.
1. remove the "Add Rule" button
   ```js
   [showAddButton]="false"
   ```
2. remove "Match" dropdown menu
   ```
   [showMatchModes]="false"
   ```
3. set the match mode to always be "Match any"
   ```
   operator="or"
   ```

See the code references:

- PrimeNG 16.9.1
  - https://github.com/primefaces/primeng/blob/16.9.1/src/app/components/table/table.ts#L4901-L5675
  - https://github.com/primefaces/primeng/blob/16.9.1/src/app/components/api/filteroperator.ts
- PrimeNG 17.0.0
  - https://github.com/primefaces/primeng/blob/17.0.0/src/app/components/table/table.ts#L4906-L5705
  - https://github.com/primefaces/primeng/blob/17.0.0/src/app/components/api/filteroperator.ts

**Reply by vokic** [reply]
Thank you! I have actually found extensive documentation for **p-columnFilter** on it on [Geeks for Geeks website](https://www.geeksforgeeks.org/angular-primeng-table-columnfilter-properties/)

Angular PrimeNG Table ColumnFilter Properties:

`field (string)` It is the name of the property of the data to filter.
`type (string)` It is the data type to display a suitable filter and matches. The default value is text.
`display (string)` It is the display mode, valid values are row for a separate row and menu for filter menu. The default value is row.
`showMenu (boolean)` It is used to set whether to show the overlay menu. The default value is true.
`matchMode (string)` It is the default match mode.
`operator (string)` It is the default operator, valid values are “and” and “or“. The default value is and.
`showOperator (boolean)` It specifies if the operator selection is displayed. The default value is true.
`showClearButton (boolean)` It specifies whether to display the clear button. The default value is true.
`showApplyButton (boolean)` It specifies whether to display the apply button. The default value is true.
`showMatchModes (boolean)` It specifies whether to display the match modes dropdown. The default value is true.
`showAddButton (boolean)` It specifies whether to display the add a rule button. The default value is true.
`showButtons (boolean)` It specifies whether to display the spinner buttons in “numeric” mode. The default value is true.
`hideOnClear (boolean)` It specifies whether to hide the overlay on clear. The default value is false.
`placeholder (string)` It is the placeholder content to display on input.
`matchModeOptions (SelectItem[])` It is the match mode option to override the global setting.
`maxConstraints (number)` It is used to set the number of maximum constraints. The default value is 2.
`minFractionDigits (number)` It is the minimum number of fraction digits to use for the numeric type.
`maxFractionDigits (number)` It is the maximum number of fraction digits to use for the numeric type. 
`prefix (string)` It is the text to display before the value for the numeric type.
`suffix (string)` It is the text to display after the value for the numeric type. The default value is decimal.
`locale (string)` It is the locale to be used in formatting for numeric type.
`localeMatcher (string)` It is the locale-matching algorithm for numeric type. The default value is the best fit. 
`currency (string)` It is the currency to use in currency formatting for the numeric type. 
`currencyDisplay (string)` It is used to set the currency to display in currency formatting for the numeric type. The default value is the symbol.
`useGrouping (boolean)` It is used to set whether to use grouping separators for “numeric” type, such as thousands of separators or thousand/lakh/crore separators.

**Reply by ThoSap** [reply]
@cetincakiroglu is it planned to add the documentation for the `<p-columnFilter/>` in PrimeNG v17?
Except the very basic **Filter Row** and **Filter Menu** examples the ColumnFilter is not documented at all, for example the input properties above can be set or that there exists a custom `pTemplate="filter"` template outlet.
For example, try to search for `showOperator` or `showMatchModes` in the current PrimeNG Table API documentation, you will find nothing.

See the code:
https://github.com/primefaces/primeng/blob/17.12.0/src/app/components/table/table.ts#L4944-L5862

See my GitHub issue:
https://github.com/primefaces/primeng/issues/13434

**Reply by cetincakiroglu** [reply]
Hi @ThoSap,

We'll revise and update all documentation including columnFilter. Thank you for the feedback, I've found that only the templating options are listed in columnFilter documentation. Thanks a lot for raising the issue!
